### PR TITLE
Private Events Project

### PR DESCRIPTION
**Pull request by necmigunduz**
In this project, we did;

- Create user, event, and event_attendee models and, relevant views, relevant controllers,
- Establish the many-to-many associations between the three models.

### CODE REVIEW

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Necmi Gunduz & Saima Anis ur Rahman** 
Good job with your first submission for this project :muscle:  :muscle:
You have defined the proper associations between the models :clap: :clap: :clap:  
There are some details to improve before you can go ahead:

- [ ] Remove commented code in events_controller.rb
- [ ] Remove commented code uin users_controller.rb
- [ ] Remove commented code in sign_in.html.erb
- [ ] The next error appears when I try to sign in:



- [ ] In the method "show" of users_controllers, use the associations you define in the models to get the events of the current users, attendees of the current user, etc.
 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    redirect_to users_show_path
+  end
+
+  def log_out
+    reset_session
+    redirect_to users_sign_in_path
+  end
+
+  def show
+    # @event = Event.joins(:attendees).where("attendee_id=#{session[:current_user_id]}")
```
Comment (acushlakoncept) [inline app/controllers/users_controller.rb]: It’s better to delete **commented piece of codes**. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))

File: app/views/events/show.html.erb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+
+
+<%#= @event.title %> <br />
+<%#= @event.description %> <br />
+<%#= @event.creator_id %> <br />
+<%#= @user.user_name %> <br />
```
Comment (acushlakoncept) [inline app/views/events/show.html.erb]: It’s better to delete **commented piece of codes**. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))

File: app/views/events/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,63 @@
+<% if logged_in? %>
+<h3><%= session[:current_user_name] %> is logged in</h3>
+
+<h2>Previous Events </h2>
```
Comment (acushlakoncept) [inline app/views/events/index.html.erb]: there should be absolutely NO logic in views.

File: app/views/events/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        <th>Event id:</th>
+        <th>Event Title:</th>
+        <th>Event Description:</th>
+        <th>Event Date</th>
+        <th>Event Location</th> 
+        <th>Creator id</th>
+        <% if logged_in? %>
+        <th>View Attendees</th> 
+        <% end %>
+    </tr>
```
Comment (acushlakoncept) [inline app/views/events/index.html.erb:16]: there should be absolutely NO logic in views.

File: app/views/events/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        <th>Event id:</th>
+        <th>Event Title:</th>
+        <th>Event Description:</th>
+        <th>Event Date</th>
+        <th>Event Location</th> 
+        <th>Creator id</th> 
+        <% if logged_in? %>
+        <th>View Attendees</th> 
+        <% end %>
+    </tr>
```
Comment (acushlakoncept) [inline app/views/events/index.html.erb]: there should be absolutely NO logic in views.

File: app/views/events/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,29 @@
+<% if logged_in? %>
+<h3><%= session[:current_user_name] %> is logged in</h3>
+
+<h3>Create a new event</h3>
```
Comment (acushlakoncept) [inline app/views/events/new.html.erb]: there should be absolutely NO logic in views.

File: app/views/layouts/application.html.erb [hunk 7]
```diff
@@ -11,6 +11,30 @@
...
+  <% if logged_in? %>
+              <p>
+                <%= link_to "Logout", users_log_out_path, method: :delete %>
+              </p>
+              <p>
+                <%= link_to "All Events", events_index_path %>
+              </p>
+              <p>
+                <%= link_to "Create an Event", events_new_path %>
+              </p>
```
Comment (acushlakoncept) [inline app/views/layouts/application.html.erb:38]: there should be absolutely NO logic in views.

File: app/views/users/show.html.erb [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+      <% if event.date < Date.today %>
+      <tr>       
+        <td><%= event.title %></td>
+        <td><%= event.description %></td>
+        <td><%= event.date %></td>
+        <td><%= event.location %></td>
+      </tr>
+      
+       <% end %>
+    <% end %>
```
Comment (acushlakoncept) [inline app/views/users/show.html.erb:24]: there should be absolutely NO logic in views.

File: app/views/users/show.html.erb [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+</table>
+<% else %>
```
Comment (acushlakoncept) [inline app/views/users/show.html.erb]: there should be absolutely NO logic in views.

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Hello, @necmigunduz and @saimaAnis21, Well done so far.  

**I would recommend always strive to care, even it's a little about UI, I know it might seem like a waste of time for a backend application, but it helps us perfect certain skills as we move along the curriculum and gives our project a professional look, you never can tell who will go through our GitHub portfolio for a job background check.**

There are still some changes that are necessary for this project to be approved.
Let us quickly add some finishing touches and request another review.


- [ ] I created events, yet non showed up on the events page, kindly fix this.

- [ ] Kind add validations to your models, I could sign up and log in with empty fields and I can also create an event with empty fields, I should not be able to do those.

- [ ] I noticed there are lots of logic in your views, there should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it.  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright)
I have gone ahead and made some comments on your codebase:
<hr>

**Testing**

- [ ] Your models should have unit tests for validations and associations (Rspec) 
([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/unit_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))

 - [ ] Your authentication should have has integration tests (Rspec + Capybara) 
([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))

 - [ ]  Your event managment features have integration tests (Rspec + Capybara) 
([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
<hr>
- [ ] Do not forget to add `yarn` to your getting started instructions, perfect after the `bundle install`. I couldn't run the project without running `yarn`

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla) or reply to this thread, if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

File: app/views/users/show.html.erb (hunk 8 above)
Comment (saimaAnis21) [reply app/views/users/show.html.erb:24]: For this particular logic in view, it was a requirement in the Odin project to separate upcoming and past events using logic in the views
image

Comment (saimaAnis21) [reply app/views/users/show.html.erb:24]: https://github.com/saimaAnis21/private-events/pull/1#discussion_r585404277

File: app/views/users/show.html.erb (hunk 9 above)
Comment (saimaAnis21) [reply app/views/users/show.html.erb]: https://github.com/saimaAnis21/private-events/pull/1#discussion_r585404277

Comment (acushlakoncept) [reply app/views/users/show.html.erb]: @saimaAnis21 that is why you have this
```
scope :upcoming, -> { where('date >= ?', DateTime.now) }
  scope :past, -> { where('date < ? ', DateTime.now) }
```
So on the events index action, you simply call `@pass_events = Events.past` and `@upcoming_events = Events.upcoming`

Comment (saimaAnis21) [reply app/views/users/show.html.erb]: If you look carefully, you will see that the logic is in the user/show page(as described in the odin project) not on the events/index page. The scope you have mentioned above is used in the events controller and I have done that already in my project. However in the users/show page it is mentioned specifically in the Odin project that the logic should be in the users show view page

Comment (acushlakoncept) [reply app/views/users/show.html.erb]: Then you use `current_user.events.past and current _user.events.upcoming`

File: app/views/users/show.html.erb (hunk 8 above)
Comment (acushlakoncept) [reply app/views/users/show.html.erb:24]: Oh! I see

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Both of you have done well on this project. Great job on adding the test 👏 
Let's make a few changes, and we will be done.

- [x] Kindly try to add a bit of spacing, everything is all joined together. You can barely tell event description from title.
Screen Shot 2021-03-02 at 6 30 04 PM

- [ ] A user should be able to attend an event, Kindly add the attend event button.

- [ ] Do not forget to add `yarn` to your getting started instructions, perfect after the `bundle install`. I couldn't run the project without running `yarn`

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla) or reply to this thread, if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by umairarshadbutt — APPROVED** [review]
# **Status: PROJECT APPROVED** :champagne:
Excellent job on this project guys. :fireworks: Keep it up.
##
**All the required tasks have been fulfilled in this project. I will go ahead and mark this as complete.**
- Congratulations on the splendid job @saimaAnis21 and @necmigunduz ! :+1: 
- You can now merge to master and submit the project completion form :rocket: . I wish you all the very best in your next tasks :1st_place_medal:
##
**If you have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

**Pull request merged**